--- test-suite/standalone/Makefile.in-orig	2022-02-16 10:30:47.143871362 -0800
+++ test-suite/standalone/Makefile.in	2022-02-16 10:33:47.834690125 -0800
@@ -106,7 +106,8 @@
 	test-scm-c-bind-keyword-arguments$(EXEEXT) \
 	test-srfi-4$(EXEEXT) $(am__append_6) $(am__EXEEXT_1) \
 	test-smob-mark$(EXEEXT) test-smob-mark-race$(EXEEXT) \
-	test-stack-overflow test-out-of-memory
+	test-stack-overflow 
+XFAIL_TESTS = test-out-of-memory
 check_PROGRAMS = test-num2integral$(EXEEXT) test-round$(EXEEXT) \
 	test-foreign-object-c$(EXEEXT) test-list$(EXEEXT) \
 	test-unwind$(EXEEXT) test-conversion$(EXEEXT) \
